### ISSUE TITLE
js_of_ocaml API and js_gen_api interop. improvements

### ISSUE THREAD

**Issue by dbuenzli** [dbuenzli/brr#2]
The only problem is that each of the libs uses `Callback.register_exception` to register the exception used to represent JavaScript exceptions in OCaml.

**Comment by hhugo** [comment]
I'm not sure how to fix this one. Any idea ?

**Comment by dbuenzli** [comment]
For reference here's where this happens in 
* [`Brr`](https://github.com/dbuenzli/brr/blob/41580885f40bfd184c3d8e5be2ddd56b0712b411/src/jv.ml#L227-L231)
* [`js_gen_api`](https://github.com/LexiFi/gen_js_api/blob/377d58632d4b34aaf54a3e2f3c80741cc1ab2044/lib/ojs_exn.ml)
* [`js_of_ocaml`](https://github.com/ocsigen/js_of_ocaml/blob/cde86b2cba215344a6f81b1db28b922b80c7b55a/lib/js_of_ocaml/js.ml#L672)

I'm not sure I have a good answer for this. I think it's the only thing were we can't rely on external primitives or `Obj.magic`. 

Maybe the solution is for `js_of_ocaml` to provide a runtime OCaml library against which `js_of_ocaml` *must* link (this could also be where the cstubs for primitives are added to avoid the `-no-check-prims` in `js_gen_api` and `brr`).

This OCaml runtime library should only have this module:
```ocaml
module Js_error : sig
  type t 
  exception Exn of t
end = struct
   type t (* Voluntarily left abstract, this is a javascript object. *)
   exception Exn of t 
   let () = Callback.register_exception "jsError" (Exn (obj [||]))
end
```
Each of the API can then simply expose its own exception implemented as being equal to `Js_error.Exn`. For example the [current Brr API](https://erratique.ch/software/brr/doc/Jv/index.html#exns) could just change its signature to: 

```ocaml
module Error : sig
   type t = Js_error.t (* <--- this is the only visible external change *)
   val name : t -> Jstr.t
   ...
end
exception Error of Error.t
```
implemented as: 
```ocaml
module Error = struct
   type t = Js_error.t
   let name e = to_jstr (get (Obj.magic e) "name")
   ...
end
exception Error = Js_error.Exn
```

**Comment by gasche** [comment]
Would it be reasonable to expect a shared library that exposes (and registers) Javascript exceptions?

If I understand correctly, @dbuenzli here describes a "minimal" approach, with binding authors adding extra features (`name`, etc.) by unsafe methods. I wonder if it is possible to have a "shared support library" approach where the useful extra features are implemented in the shared library directly -- assuming that what binding libraries need/want is not so different on each side.

**Comment by dbuenzli** [comment]
> Would it be reasonable to expect a shared library that exposes (and registers) Javascript exceptions?

No. A JavaScript exception is just an OCaml exception holding a JavaScript *object* and the way these are represented and exposed by `brr` and `js_gen_api` is quite different from `js_of_ocaml`.

> If I understand correctly, @dbuenzli here describes a "minimal" approach, with binding authors adding extra features (`name`, etc.) by unsafe methods.

I would't say *binding authors*, I would say *FFI API authors*. That's fine, there shouldn't be dozens of them.

Really the only thing that is needed to share the above 8 lines modules.

**Comment by mlasson** [comment]
> >   If I understand correctly, @dbuenzli here describes a "minimal" approach, with binding authors adding extra features (name, etc.) by unsafe methods.
> 
> I wouldn't say binding authors, I would say FFI API authors. That's fine, there shouldn't be dozens of them.

FWIW we would be happy to use in `gen_js_api` a shared library to register the exception (even more if this library contains the c-stubs !). 

If this library contain a bit more (something a bit like your "jv" module) to completly replace our "ojs" module that could be even better for interop !

**Comment by hhugo** [comment]
@dbuenzli, would you like to open an issue on jsoo ?

**Comment by dbuenzli** [comment]
Sorry I missed your suggestion done in https://github.com/ocsigen/js_of_ocaml/issues/1146

**Comment by dbuenzli** [comment]
Solved by 6a1745828280f6e8f5f1e and 0b48f2871eb5ee6. 

Thanks Hugo for the nice work in `js_of_ocaml`.
